aten/src/ATen/native/UpSample.h

```diff
@@ -0,0 +1,249 @@
+  AT_CHECK(
+      data.dim() == dim && data.size(dim_size) == size,
+      "Expected tensor of dimension ",
+      dim,
+      " and tensor.size[",
+      dim_size,
+      "] == ",
+      size,
+      " but got: dimension ",
+      data.dim(),
+      " and tensor.size[",
+      dim_size,
+      "] = ",
+      data.size(dim_size));
+}
+
+static inline void upsample_1d_shape_check(
+    const Tensor& data,
+    int64_t type_check,
+    int64_t nbatch,
+    int64_t nchannels,
+    int64_t input_width,
+    int64_t output_width) {
+  AT_CHECK(
+      input_width > 0 && output_width > 0,
+      "Input and output sizes should be greater than 0, but got input (W: ",
+      input_width,
+      ") and output (W: ",
+      output_width,
+      ")");
+
+  if (type_check == 0) {
+    AT_CHECK(
+        !data.numel() == 0 && data.dim() == 3,
+        "Non-empty 3D data tensor expected but got: ",
+        data.dim(),
+        "D");
+  } else if (type_check == 1) {
+    check_dim_size(data, 3, 0, nbatch);
+    check_dim_size(data, 3, 1, nchannels);
+    check_dim_size(data, 3, 2, output_width);
+  }
+}
+
+static inline void upsample_2d_shape_check(
+    const Tensor& data,
+    int64_t type_check,
+    int64_t nbatch,
+    int64_t nchannels,
+    int64_t input_height,
+    int64_t input_width,
+    int64_t output_height,
+    int64_t output_width) {
+  AT_CHECK(
+      input_height > 0 && input_width > 0 && output_height > 0 &&
+          output_width > 0,
+      "Input and output sizes should be greater than 0,"
+      " but got input (H: ",
+      input_height,
+      ", W: ",
+      input_width,
+      ") output (H: ",
+      output_height,
+      ", W: ",
+      output_width,
+      ")");
+
+  if (type_check == 0) {
+    AT_CHECK(
+        !data.numel() == 0 && data.dim() == 4,
+        "Non-empty 4D data tensor expected but got: ",
+        data.dim(),
+        "D");
+  } else if (type_check == 1) {
+    check_dim_size(data, 4, 0, nbatch);
+    check_dim_size(data, 4, 1, nchannels);
+    check_dim_size(data, 4, 2, output_height);
+    check_dim_size(data, 4, 3, output_width);
+  }
+}
+
+static inline void upsample_3d_shape_check(
+    const Tensor& data,
+    int64_t type_check,
+    int64_t nbatch,
+    int64_t nchannels,
+    int64_t input_depth,
+    int64_t input_height,
+    int64_t input_width,
+    int64_t output_depth,
+    int64_t output_height,
+    int64_t output_width) {
+  AT_CHECK(
+      input_depth > 0 && input_height > 0 && input_width > 0 &&
+          output_depth > 0 && output_height > 0 && output_width > 0,
+      "Input and output sizes should be greater than 0, but got input (D: ",
+      input_depth,
+      ", H: ",
+      input_height,
+      ", W: ",
+      input_width,
+      ") output (D: ",
+      output_depth,
+      ", H: ",
+      output_height,
+      ", W: ",
+      output_width,
+      ")");
+
+  if (type_check == 0) {
+    AT_CHECK(
+        data.dim() == 5, "5D data tensor expected but got: ", data.dim(), "D");
```

Comment:
Arguably a bug in the old code, but we probably should be checking that it's non-empty here too.

Reply:
sounds good.